### ISSUE TITLE
CI fails with latest version of xarray

### ISSUE THREAD

**Issue by xylar** [MPAS-Dev/MPAS-Analysis#95]
CI has started to fail despite no changes to MPAS-Analysis itself.  The error would appear to be because of an update to `xarray`.  Here is the result of pytest:

```
$ pytest
============================= test session starts ==============================
platform linux2 -- Python 2.7.13, pytest-3.0.5, py-1.4.32, pluggy-0.4.0
rootdir: /home/xylar/code/mpas-work/analysis/mpas_analysis_repo, inifile: 
collected 17 items 

mpas_analysis/test/test_date.py .
mpas_analysis/test/test_interpolate.py ....
mpas_analysis/test/test_io_utility.py .
mpas_analysis/test/test_mpas_config_parser.py .
mpas_analysis/test/test_mpas_xarray.py ....F..
mpas_analysis/test/test_namelist_streams_interface.py ...

=================================== FAILURES ===================================
__________________________ TestNamelist.test_selvals ___________________________

self = <mpas_analysis.test.test_mpas_xarray.TestNamelist testMethod=test_selvals>

    def test_selvals(self):
        fileName = str(self.datadir.join('example_jan.nc'))
        timestr = 'time_avg_daysSinceStartOfSim'
        varList = \
            ['time_avg_avgValueWithinOceanLayerRegion_avgLayerTemperature',
             'refBottomDepth']
    
        selvals = {'nVertLevels': 0}
        ds = xr.open_mfdataset(
            fileName,
            preprocess=lambda x: mpas_xarray.preprocess_mpas(x,
                                                             timestr=timestr,
                                                             onlyvars=varList,
                                                             selvals=selvals,
                                                             yearoffset=1850))
        self.assertEqual(ds.data_vars.keys(), varList)
        self.assertEqual(ds[varList[0]].shape, (1, 7))
>       self.assertEqual(ds['nVertLevels'].shape, ())

/home/xylar/code/mpas-work/analysis/mpas_analysis_repo/mpas_analysis/test/test_mpas_xarray.py:136: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/xylar/anaconda2/envs/mpas_analysis/lib/python2.7/site-packages/xarray/core/dataset.py:698: in __getitem__
    return self._construct_dataarray(key)
/home/xylar/anaconda2/envs/mpas_analysis/lib/python2.7/site-packages/xarray/core/dataset.py:642: in _construct_dataarray
    self._variables, name, self._level_coords, self.dims)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

variables = OrderedDict([(u'time_avg_avgValueWithinOceanLayerRegion_avgLayerTemperature', ...(Time: 1)>
array(['1855-01-13T12:24:13.593599000Z'], dtype='datetime64[ns]'))])
key = 'nVertLevels', level_vars = OrderedDict()
dim_sizes = Frozen(SortedKeysDict({u'nOceanRegionsTmp': 7, u'Time': 1}))

    def _get_virtual_variable(variables, key, level_vars=None, dim_sizes=None):
        """Get a virtual variable (e.g., 'time.year' or a MultiIndex level)
        from a dict of xarray.Variable objects (if possible)
        """
        if level_vars is None:
            level_vars = {}
        if dim_sizes is None:
            dim_sizes = {}
    
        if key in dim_sizes:
            data = pd.Index(range(dim_sizes[key]), name=key)
            variable = IndexVariable((key,), data)
            return key, key, variable
    
        if not isinstance(key, basestring):
            raise KeyError(key)
    
        split_key = key.split('.', 1)
        if len(split_key) == 2:
            ref_name, var_name = split_key
        elif len(split_key) == 1:
            ref_name, var_name = key, None
        else:
            raise KeyError(key)
    
        if ref_name in level_vars:
            dim_var = variables[level_vars[ref_name]]
            ref_var = dim_var.to_index_variable().get_level_variable(ref_name)
        else:
>           ref_var = variables[ref_name]
E           KeyError: 'nVertLevels'

/home/xylar/anaconda2/envs/mpas_analysis/lib/python2.7/site-packages/xarray/core/dataset.py:71: KeyError
===================== 1 failed, 16 passed in 0.87 seconds ======================
```

**Comment by xylar** [comment]
I did a test on my laptop with xarray 0.8.2.  Then, I did a `conda update --all` and got the following updates:
```
 $ conda update --all
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ....
Solving package specifications: .........

Package plan for installation in environment /home/xylar/anaconda2/envs/mpas_analysis:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.0.2k             |                0         3.2 MB
    qtpy-1.2.1                 |           py27_0          28 KB
    isort-4.2.5                |           py27_0          51 KB
    qtawesome-0.4.3            |           py27_0         157 KB
    pylint-1.6.4               |           py27_1         316 KB
    xarray-0.9.1               |           py27_0         443 KB
    ------------------------------------------------------------
                                           Total:         4.1 MB

The following NEW packages will be INSTALLED:

    isort:     4.2.5-py27_0

The following packages will be UPDATED:

    openssl:   1.0.2j-0     --> 1.0.2k-0    
    pylint:    1.6.4-py27_0 --> 1.6.4-py27_1
    qtawesome: 0.4.1-py27_0 --> 0.4.3-py27_0
    qtpy:      1.2.0-py27_0 --> 1.2.1-py27_0
    xarray:    0.8.2-py27_0 --> 0.9.1-py27_0
```
Then, I get the same error as the CI.

**Comment by xylar** [comment]
It appears that `xarray` has changed the behavior when a key value is given that doesn't exist.  The new behavior is to raise a `KeyError`, which would seem to be the correct behavior to me.  Apparently the old behavior was to create an empty list, which seems like the wrong behavior.  This seems like an easy fix -- the test simply needs to be updated to follow the new (correct) behavior -- it should check to make sure a `KeyError` was raised.

https://github.com/MPAS-Dev/MPAS-Analysis/blob/develop/mpas_analysis/test/test_mpas_xarray.py#L136

I will submit a PR shortly.

**Comment by xylar** [comment]
Ah, the error is a little more complicated than I realized.  It seems you used to be able to do `ds[dim]` to get a dimension but it seems like that syntax is no longer supported.  I'm looking into it...

**Comment by xylar** [comment]
It seems like:
1) `ds[dimName]` no longer works, only `ds[varArrayName]`.
2) it seems that selecting a single element in a dimension (e.g. passing `selvals = {'nVertLevels': 0}` to `mpas_xarray.preprocess_mpas`) causes that dimension to go away completely, which was clearly not the behavior in previous versions.

This would seem to mean that information is lost. @pwolfram, do you think this is a bug?  I would tend to think it might be or it might at least be an unintended side effect of some other change.  Could you look into it or ask the xarray folks if this was intentional? Thanks!

**Comment by xylar** [comment]
Reading more, this behavior seems to be intentional. 
http://xarray.pydata.org/en/stable/whats-new.html
It seems like what they call "Dimensions without coordinates" are not handled the same way they used to be.  `nVertLevels` would be an example of this in our test.

@pwolfram, I'm not quite sure how we want to update the test.  The following lines in `test_mpas_xarray.test_selvals` are no longer valid:
```
        self.assertEqual(ds['nVertLevels'].shape, ())
        self.assertApproxEqual(ds['nVertLevels'],
                               selvals['nVertLevels'])
```
You can get something like the behavior in the first by passing `selvals={'nVertLevels': [10]}` with a single element list instead of a single index.  But the dataset doesn't remember that `nVertLevels` was indexed at level 10 after that.  If you do `ds['nVertLevels'][0]` you get 0, not 10.  So there's no way to recover the information that the second test is trying to get.

Hopefully we're not counting on this capability that has been removed elsewhere in our code.  It isn't a very intuitive way of using xarray's indexing anyway so it's maybe for the best that it's not supported.  That is, `nVertLevels=10` isn't a very intuitive way of saying that you want level numer 10.  We should presumably be using isel to index "Dimensions without coordinates" like `nVertLevels`.

**Comment by milenaveneziani** [comment]
I am afraid this will break the code in a few places, not just the testing.
For sure we have it in various places in replicate_cycle, when we do calculations on dsnew['Time'], for example. Now, perhaps replicate_cycle will soon be completely different, but we need to check the other scripts. I can do it, but an extra set of eyes would be better.

**Comment by xylar** [comment]
Should be fixed by #96.

**Comment by xylar** [comment]
@milenaveneziani, see my comments in #97.
